sql/core/src/main/scala/org/apache/spark/sql/streaming/ui/StreamingQueryStatusListener.scala

```diff
@@ -18,104 +18,161 @@
+      false,
+      querySummary.exception
+    ))
   }
 }
 
 /**
  * This class contains all message related to UI display, each instance corresponds to a single
  * [[org.apache.spark.sql.streaming.StreamingQuery]].
  */
-private[ui] class StreamingQueryUIData(
+private[sql] class StreamingQueryUIData(
     val name: String,
     val id: UUID,
-    val runId: UUID,
-    val startTimestamp: Long) {
+    @KVIndexParam val runId: UUID,
+    val progressIdQueue: Queue[String],
+    val startTimestamp: Long,
+    val isActive: Boolean,
+    val exception: Option[String]) {
 
-  /** Holds the most recent query progress updates. */
-  private val progressBuffer = new mutable.Queue[StreamingQueryProgress]()
+  private var storeOption: Option[ElementTrackingStore] = None
 
-  private var _isActive = true
-  private var _exception: Option[String] = None
+  @JsonIgnore @KVIndex("startTimestamp")
+  private def startTimestampIndex: Long = startTimestamp
 
-  def isActive: Boolean = synchronized { _isActive }
-
-  def exception: Option[String] = synchronized { _exception }
-
-  def queryTerminated(event: StreamingQueryListener.QueryTerminatedEvent): Unit = synchronized {
-    _isActive = false
-    _exception = event.exception
+  def recentProgress: Array[StreamingQueryProgress] = {
+    storeOption.map { store => progressIdQueue.map { uniqueId =>
+      store.read(classOf[StreamingQueryProgressWrapper], uniqueId).progress
+    }.toArray }.getOrElse(Array.empty)
   }
 
-  def updateProcess(
-      newProgress: StreamingQueryProgress, retentionNum: Int): Unit = progressBuffer.synchronized {
-    progressBuffer += newProgress
-    while (progressBuffer.length >= retentionNum) {
-      progressBuffer.dequeue()
+  def lastProgress: StreamingQueryProgress = {
+    if (progressIdQueue.nonEmpty) {
+      storeOption.map(_.read(classOf[StreamingQueryProgressWrapper], progressIdQueue.last).progress)
+        .orNull
+    } else {
+      null
     }
   }
 
-  def recentProgress: Array[StreamingQueryProgress] = progressBuffer.synchronized {
-    progressBuffer.toArray
+  def setKVStore(store: ElementTrackingStore): StreamingQueryUIData = {
+    storeOption = Option(store)
+    this
   }
+}
 
-  def lastProgress: StreamingQueryProgress = progressBuffer.synchronized {
-    progressBuffer.lastOption.orNull
+private[sql] class StreamingQueryProgressWrapper(val progress: StreamingQueryProgress) {
+  @KVIndexParam("batchId") val batchId: Long = progress.batchId
+  @KVIndexParam("runId") val runId: String = progress.runId.toString
+
+  @JsonIgnore @KVIndex
+  def uniqueId: String = getUniqueId(progress.runId, progress.batchId, progress.timestamp)
+}
+
+private[sql] object StreamingQueryProgressWrapper {
+  /**
+   * Adding `timestamp` into unique id to support reporting `empty` query progress
+   * when no data comes.
+   */
+  def getUniqueId(
+      runId: UUID,
+      batchId: Long,
+      timestamp: String): String = {
+    s"${runId}_${batchId}_$timestamp"
```

Comment:
The unique id of streaming query process info in kvstore. It contains `runId`, `batchId` and `timestamp`. We use `timestamp` to distinguish the empty process report which has the same `batchId`